Add support for more types passed to `context.json` helper

wyze opened the pull request:
Close #349

Expand the types the `context.json` helper accepts for the `body` argument so it is not just `Record<string, any>`. It should accept any type that would result in a successful `JSON.stringify` call.

wyze committed:
Add support for more types passed to `context.json` helper

Close #349

msutkowski commented on src/context/json.ts:
```diff
@@ -1,11 +1,17 @@
+type JsonValue = string | boolean | number | JsonRecord | JsonArray | null
+
 /**
  * Sets the given Object as the JSON body of the response.
```
```suggestion
 * Sets the given value as the JSON body of the response.
```

msutkowski commented on src/context/json.ts:
```diff
@@ -1,11 +1,17 @@
 
+type JsonArray = (JsonValue | JsonRecord)[]
+type JsonRecord = Record<string, unknown>
+type JsonValue = string | boolean | number | JsonRecord | JsonArray | null
```
This looks pretty good! I like the idea of being this strict, but it seems like we should consider matching `JSON.stringify`'s type of `any` here. 

The only gotcha I can think of here with these types as-is is that this would potentially force a TypeScript user to transform `Date`s before being passed in. I don't personally think that's really a problem but wanted to bring it up for @kettanaito's review.

wyze replied:
Oops! Definitely should support Date since the method converts it to a string.

marcosvega91 replied:
I agree with this because is what `JSON.stringify` does 

```js
/**
* Converts a JavaScript value to a JavaScript Object Notation (JSON) string.
* @param value A JavaScript value, usually an object or array, to be converted.
* @param replacer An array of strings and numbers that acts as a approved list for selecting the object properties that will be stringified.
* @param space Adds indentation, white space, and line break characters to the return-value JSON text to make it easier to read.
*/
stringify(value: any, replacer?: (number | string)[] | null, space?: string | number): string;
```

kettanaito replied:
I agree with @msutkowski on typing the value as `any`, just as `JSON.stringify` has. That leaves the type annotation aligned with the underlying stringification, and reduces the cost of maintaining it in the future. If an invalid value is passed, an error will be raised when stringifying it. It's the end developer's responsibility to pass a valid JSON to `ctx.json()`.

msutkowski reviewed (commented):
Awesome work! Just adding a few comments for further review.

kettanaito commented on src/context/json.test.ts:
```diff
@@ -3,18 +3,28 @@ import { response } from '../response'
+
+      const bool = response(json(true))
+
+      expect(bool).toHaveProperty('body', 'true')
```
Could you please add a `Date` usage scenario here?

kettanaito commented on src/context/json.test.ts:
```diff
@@ -3,18 +3,28 @@ import { response } from '../response'
-      expect(result).toHaveProperty('body', `{"firstName":"John"}`)
+      const object = response(json({ firstName: 'John' }))
+
+      expect(object).toHaveProperty('body', `{"firstName":"John"}`)
```
I would truncate these assertions to call `response()` within the `expect`. It doesn't read right, as calling the `response` function always returns a response, not object/array/etc.

```js
expect(response(json({ firstName: 'John' }))).toHaveProperty('body', `{"firstName":"John"}`)
```

kettanaito commented on src/context/json.ts:
```diff
@@ -1,11 +1,17 @@
+ * res(json([1, '2', false, { ok: true }]))
  */
-export const json = (body: Record<string, any>): ResponseTransformer => {
+export const json = (body: JsonValue): ResponseTransformer => {
```
Based on our discussion, let's annotate it as `body: any`, similar to `JSON.stringify` argument.

kettanaito reviewed (changes requested):
Thank you for preparing these improvements, @wyze! 

It looks great, I've had a few comments on how we can shorten our test scenarios, and a point about type annotating the JSON value. Could you please take a look and address them? Thanks.

wyze committed:
Respond to PR feedback

- Update comment to reflect usage
- Update type to match `JSON.stringify`
- Cleanup tests and add a date assertion

kettanaito reviewed (approved):
Looks great! Thank you for your involvement, @wyze. Welcome to contributors! 🎉
